### ISSUE TITLE
FFT analyzer

### ISSUE THREAD

**Issue by shiftyscales** [Frooxius/NeosPublic#1207]
Requested by various users, a FFT analyzer to enable sound visualization.

Enverex has provided a [potentially useful link](https://www.codeproject.com/Articles/1095473/Comparison-of-FFT-Implementations-for-NET) benchmarking various implementations.

The article's conclusion was that '[FFTW](http://www.fftw.org/)' was the best available (as of the time of the article's writing in 2016), although the [FFTW library](https://github.com/FFTW/fftw3) is coded in C, and would need a [wrapper](https://github.com/tszalay/FFTWSharp).

**Comment by inzanesonikku** [comment]
YES PLEASE I CAN NOT TELL YOU ENOUGH HOW MUCH I WANT THIS

**Comment by Frooxius** [comment]
This is planned, but I have to think more about how it will fit architecturally into Neos. The library itself isn't a big of a problem (although it's unlikely I'll use native one for the added complexity), even the audio decoding library we're using right now has FFT implementation.

The main problem is that the result size can be variable/configurable, which means you'd need collections to access the analysis results. The proper result is also actually complex numbers, which Neos doesn't support natively right now. It could be dumbed down to just floats, but at the same time that leads to information loss.

**Comment by RileyGuy** [comment]
Couldn't dynamic variables be used to grab the FFT values? I feel like you could mitigate precision loss by just making them doubles instead of floats.

**Comment by Frooxius** [comment]
No, dynamic variables don't have anything to do with this, that's something you'd use separately on top of this system. It needs to be properly implemented in the data model first.

The issue isn't precision loss, it's information loss. Doubles are still real numbers, not complex numbers. The problem is that you throw the imaginary part of the complex number away when you do that.

**Comment by DoubleStyx** [comment]
What if we use the Hartley transform instead? It both accepts and returns real numbers. Also, we technically don't _need_ collections; we could specify a certain frequency to sample from and interpolate between frequency-space array elements.

**Comment by BlueCyro** [comment]
> What if we use the Hartley transform instead? It both accepts and returns real numbers. Also, we technically don't _need_ collections; we could specify a certain frequency to sample from and interpolate between frequency-space array elements.

CSCore already has a handy FFT function, we could just use that since it's already in neos.
I've got something cooking up that people may like. I'll post it here when it's finished.


> Couldn't dynamic variables be used to grab the FFT values? I feel like you could mitigate precision loss by just making them doubles instead of floats.

Hi, it's me from nearly 2 years later, this was real dumb :p

**Comment by BlueCyro** [comment]
> I've got something cooking up that people may like, I'll post it here when it's finished.

Well, it's finished!
https://github.com/RileyGuy/FourierForge
Enjoy!
